ICal recurrence support

adamfranco opened the pull request:
Hello All,

These three commits improve support for recurrence in iCal feeds by allowing multi-day weekly recurrence and by allowing for exceptions to event recurrence. Lastly, there is a fix that make UNTIL limits work correctly.

If you would like to see this support in action, our Library Hours calendar in Google Calendar makes use of both multi-day weekly recurrence and exceptions to the normal recurrence schedule:
- http://www.google.com/calendar/embed?src=middlebury.edu_ol2e145ahtv7tga6qavgcoretc@group.calendar.google.com&ctz=America/New_York&mode=week
- http://www.google.com/calendar/ical/middlebury.edu_ol2e145ahtv7tga6qavgcoretc@group.calendar.google.com/public/basic.ics

Let me know if there are any modifications needed.

Best,
Adam

adamfranco committed:
ICalRecurrenceRule now supports WEEKLY recurrence with the BYDAY attribute.

A common (and relatively simple) recurrence rule is to repeat several days
per week, every week as seen in the example below.

Example:

	BEGIN:VEVENT
	DTSTART;TZID=America/New_York:20110620T073000
	DTEND;TZID=America/New_York:20110621T000000
	RRULE:FREQ=WEEKLY;UNTIL=20110811T113000Z;BYDAY=MO,TU,WE,TH
	DTSTAMP:20110824T134544Z
	UID:o0742l784numhoobgl50sh58f4@google.com
	CREATED:20110504T185446Z
	DESCRIPTION:
	LAST-MODIFIED:20110603T034935Z
	LOCATION:
	SEQUENCE:0
	STATUS:CONFIRMED
	SUMMARY:Regular hours are 7:30 am - midnight
	TRANSP:OPAQUE
	END:VEVENT

Additional recurrence cycles such as BYMONTH, BYWEEKNO, BYMONTHDAY, and
INTERVALs are not supported by this change, but could be implemented
in a similar fashion. For more examples of recurrence cycles see:
    http://www.kanzaki.com/docs/ical/rrule.html

adamfranco committed:
ICalendar now supports exceptions to event recurrence.

VCALENDARs will often have multiple VEVENTs with the same UID. One of the
events will have an RRULE indicating a recurrence cycle. The others will
have a RECURRENCE-ID that matches the time of one of the recurrence events
and overrides the content of the recurring event for that single occurrence.

Example:

	BEGIN:VEVENT
	DTSTART;TZID=America/New_York:20110825T090000
	DTEND;TZID=America/New_York:20110825T150000
	DTSTAMP:20110824T134544Z
	UID:vmkvs3qmsmpfsnqet6pl7vlpkk@google.com
	RECURRENCE-ID;TZID=America/New_York:20110825T090000
	CREATED:20110623T172558Z
	DESCRIPTION:
	LAST-MODIFIED:20110719T124934Z
	LOCATION:
	SEQUENCE:0
	STATUS:CONFIRMED
	SUMMARY:9 am - 3 pm (early closing)
	TRANSP:OPAQUE
	END:VEVENT

	BEGIN:VEVENT
	DTSTART;TZID=America/New_York:20110815T090000
	DTEND;TZID=America/New_York:20110815T170000
	RRULE:FREQ=WEEKLY;UNTIL=20110902T130000Z;BYDAY=MO,TU,WE,TH,FR
	DTSTAMP:20110824T134544Z
	UID:vmkvs3qmsmpfsnqet6pl7vlpkk@google.com
	CREATED:20110623T172558Z
	DESCRIPTION:
	LAST-MODIFIED:20110623T172558Z
	LOCATION:
	SEQUENCE:0
	STATUS:CONFIRMED
	SUMMARY:9 am - 5 pm
	TRANSP:OPAQUE
	END:VEVENT

For more details on RECURRENCE-ID see:
	http://www.kanzaki.com/docs/ical/recurrenceId.html

adamfranco committed:
Moved ICalRecurrenceRule checking of UNTIL limits so that they work correctly.

Occurrences that were beyond the limit were being included since they were
added before the UNTIL check was performed.

akinspe commented:
Thank you so much for this. I will be analyzing it further and hopefully merging it in next week.

adamfranco committed:
Renamed ICalRecurrenceRule WEEKLY BYDAY variables for readability.

 -> ,  ->
Using  rather than the generic  makes the loop easier to read and its relationship to
 more clear.
